hs2-types/src/Hs2/Types/Type.hs

```diff
@@ -0,0 +1,24 @@
+module Hs2.Types.Type where
+
+type Cxt = Type
+
+data PromotableType
+  = ConTy Name
+  | VarTy Name
+  | TupleTy Int
+
+data Type
```

Comment:
@kcsongor @TOTBWF what would happen if we added a type lambda while we're here?

Reply:
Things get a bit tricky with type level lambdas and instances. Unification of type level lambdas is undecidable in the general case, so you either need to restrict their usage, try not to unify them, or just accept the undecidablility.

Reply:
That's helpful, thanks. In the pros column, what would it buy us much?

Reply:
I suppose it would help avoid newtypes from time to time, but beyond that I think the cons outweigh the pros. Perhaps someone else has better insight into this?

Reply:
I think full dependent types is probably outside of the scope here :) Primarily my thinkinig here I was trying to address @masaeedu's concerns that typeclasses are rigid on what parameters they let you play with. If you had type lambdas, at least in instance heads, you could shuffle around tyvars.